cmd/notification-file/main.go

```diff
@@ -0,0 +1,253 @@
+	MaxSize  int  `yaml:"max_size"`
+	MaxAge   int  `yaml:"max_age"`
+	MaxFiles int  `yaml:"max_files"`
+	Enabled  bool `yaml:"enabled"`
+	Compress bool `yaml:"compress"`
+}
+
+type FilePlugin struct {
+	PluginConfigByName map[string]PluginConfig
+}
+
+var logger hclog.Logger = hclog.New(&hclog.LoggerOptions{
+	Name:       "file-plugin",
+	Level:      hclog.LevelFromString("INFO"),
+	Output:     os.Stderr,
+	JSONFormat: true,
+})
+
+func (r *LogRotate) rotateLogs(cfg PluginConfig) {
+	// Rotate the log file
+	err := r.rotateLogFile(cfg.LogPath, r.MaxFiles)
+	if err != nil {
+		logger.Error("Failed to rotate log file", "error", err)
+	}
+	// Reopen the FileWriter
+	FileWriter.Close()
+	FileWriter, err = os.OpenFile(cfg.LogPath, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
+	if err != nil {
+		logger.Error("Failed to reopen log file", "error", err)
+	}
+	// Reset the file size
+	FileInfo, err := FileWriter.Stat()
+	if err != nil {
+		logger.Error("Failed to get file info", "error", err)
+	}
+	FileSize = FileInfo.Size()
+}
+
+func (r *LogRotate) rotateLogFile(logPath string, maxBackups int) error {
+	// Rename the current log file
+	backupPath := logPath + "." + time.Now().Format("20060102-150405")
+	err := os.Rename(logPath, backupPath)
+	if err != nil {
+		return err
+	}
+	glob := logPath + ".*"
+	if r.Compress {
+		glob = logPath + ".*.gz"
+		err = compressFile(backupPath)
+		if err != nil {
+			return err
+		}
+	}
+
+	// Remove old backups
+	files, err := filepath.Glob(glob)
+	if err != nil {
+		return err
+	}
+
+	sort.Sort(sort.Reverse(sort.StringSlice(files)))
+
+	for i, file := range files {
+		logger.Trace("Checking file", "file", file, "index", i, "maxBackups", maxBackups)
+		if i >= maxBackups {
+			logger.Trace("Removing file as over max backup count", "file", file)
+			os.Remove(file)
+		} else {
+			// Check the age of the file
+			fileInfo, err := os.Stat(file)
+			if err != nil {
+				return err
+			}
+			age := time.Since(fileInfo.ModTime()).Hours()
+			if age > float64(r.MaxAge*24) {
+				logger.Trace("Removing file as age was over configured amount", "file", file, "age", age)
+				os.Remove(file)
+			}
+		}
+	}
+
+	return nil
+}
+
+func compressFile(src string) error {
+	// Open the source file for reading
+	srcFile, err := os.Open(src)
+	if err != nil {
+		return err
+	}
+	defer srcFile.Close()
+
+	// Create the destination file
+	dstFile, err := os.Create(src + ".gz")
+	if err != nil {
+		return err
+	}
+	defer dstFile.Close()
+
+	// Create a gzip writer
+	gw := gzip.NewWriter(dstFile)
+	defer gw.Close()
+	// Read the source file and write its contents to the gzip writer
+	b, err := io.ReadAll(srcFile)
+	if err != nil {
+		return err
+	}
+	_, err = gw.Write(b)
+	if err != nil {
+		return err
+	}
+
+	// Delete the original (uncompressed) backup file
+	err = os.Remove(src)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func WriteToFileWithCtx(ctx context.Context, cfg PluginConfig, log string) error {
+	FileWriteMutex.Lock()
```

Comment:
Need to check with team as I dont even know if mutexs are even needed tested without and it was "fine", however, I dont have a high load test env and when reading the code that sends to plugin channel it inside a go routine so could be written multiple times 🤷🏻